### ISSUE TITLE
Terraform crash in s3 backend when dynamodb record is present but s3 object has been deleted

### ISSUE THREAD

**Issue by mildred** [hashicorp/terraform#16634]
### Terraform Version
0.9.6

### Crash Output

I do not have crash.log because this is running inside an automated process and the files are removed automatically. I think however that the stack trace is enough to understand what's wrong:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x3af5cba]

goroutine 1804 [running]:
github.com/hashicorp/terraform/backend/remote-state/s3.(*RemoteClient).Get(0xc4202f93b0, 0x11, 0x11, 0x18)
	/home/travis/gopath/src/github.com/squarescale/squarescale-deploy/src/github.com/hashicorp/terraform/backend/remote-state/s3/client.go:64 +0x11a
github.com/hashicorp/terraform/state/remote.(*State).RefreshState(0xc4205a55e0, 0x40eed2, 0xc4203248c0)
	/home/travis/gopath/src/github.com/squarescale/squarescale-deploy/src/github.com/hashicorp/terraform/state/remote/state.go:33 +0x37
github.com/hashicorp/terraform/state.(*BackupState).RefreshState(0xc4204cb800, 0x5221db5, 0x7)
	/home/travis/gopath/src/github.com/squarescale/squarescale-deploy/src/github.com/hashicorp/terraform/state/backup.go:21 +0x33
github.com/hashicorp/terraform/backend/local.(*Local).context(0xc42089c780, 0xc4207d6580, 0x2, 0x2, 0x3, 0xc4203248c0, 0x2)
	/home/travis/gopath/src/github.com/squarescale/squarescale-deploy/src/github.com/hashicorp/terraform/backend/local/backend_local.go:31 +0x97
github.com/hashicorp/terraform/backend/local.(*Local).opPlan(0xc42089c780, 0x7cd1880, 0xc420018400, 0xc4207d6580, 0xc4204ed0b0)
	/home/travis/gopath/src/github.com/squarescale/squarescale-deploy/src/github.com/hashicorp/terraform/backend/local/backend_plan.go:57 +0x22f
github.com/hashicorp/terraform/backend/local.(*Local).(github.com/hashicorp/terraform/backend/local.opPlan)-fm(0x7cd1880, 0xc420018400, 0xc4207d6580, 0xc4204ed0b0)
	/home/travis/gopath/src/github.com/squarescale/squarescale-deploy/src/github.com/hashicorp/terraform/backend/local/backend.go:243 +0x52
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc42089c780, 0xc42060a120, 0xc42060a110, 0x7cd1880, 0xc420018400, 0xc4207d6580, 0xc4204ed0b0)
	/home/travis/gopath/src/github.com/squarescale/squarescale-deploy/src/github.com/hashicorp/terraform/backend/local/backend.go:265 +0x9a
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/home/travis/gopath/src/github.com/squarescale/squarescale-deploy/src/github.com/hashicorp/terraform/backend/local/backend.go:266 +0x192
```

### Important Factoids

I am cleaning up tfstate in s3 but I did not necessarily cleaned up the dynamodb records. I believe the crash is caused by a dynamodb record that is left while I removed the tfstate in S3. The problem occurs in this function:

https://github.com/hashicorp/terraform/blob/85e0979c6af6a15345c9131e4478ba712bb3c21d/backend/remote-state/s3/client.go#L50-L84

Here we have the following function call that fetches the state in S3:

```go
		payload, err = c.get()
		if err != nil {
			return nil, err
		}
```

When there is no S3 state, `err` is nil and `payload` is also nil. We do not exit early and we continue the function with a nil `payload`. Now the following code is executed:

```go
		// verify that this state is what we expect
		if expected, err := c.getMD5(); err != nil {
			log.Printf("[WARNING] failed to fetch state md5: %s", err)
		} else if len(expected) > 0 && !bytes.Equal(expected, payload.MD5) {
			log.Printf("[WARNING] state md5 mismatch: expected '%x', got '%x'", expected, payload.MD5)
```

Here, `c.getMD5()` wil fetch the dynamodb record and the record is found. We have no error and `expected` is probably greater than 0. The next conditional crashes because `len(expected) > 0` is true and `!bytes.Equal(expected, payload.MD5)` crashes because `payload` is nil.

**Comment by mildred** [comment]
Looking at the code in master, the bug is still there.

**Comment by jbardin** [comment]
Duplicate of #14927, fixed by #14937
